A21-Service-Config-Error-Handling.md

```diff
@@ -0,0 +1,174 @@
+known fields, valid values are defined by the field itself. If any of the known
+fields fails in validation, the entire service config is deemed invalid.
+
+## Rationale
+
+By discarding invalid service configs instead of ignoring fields that have
+invalid values, we prevent clients from resorting to using default values which
+might be unwanted for certain systems. The clause that existing clients continue
+using older service configs prevents an invalid service config from bringing
+down the entire system, while newer clients which do not have a previously
+received valid service config fail to start, avoiding new traffic for the
+duration that the service is set with an invalid config.
+
+Examples of invalid service configs :
+
+* Badly Formatted service config
+```
+“serviceConfig” : {
+  “MethodConfig” : {
+    “Service // bad format
+  }
+```
+
+* Fields with invalid values are not allowed
+```
+“serviceConfig” : {
+  “loadBalancingPolicy” : “UnknownPolicy”
+}
+```
+
+Note that loadBalancingPolicy is deprecated in favor of loadBalancingConfig,
+which provides a list of LB policies with the client selecting the first policy
+it supports. If it doesn’t understand any of the policies, it is treated as
+invalid, and the entire service config is rejected.
+
+```
+“serviceConfig” : {
+  “loadBalancingConfig” : [
+    “UnknownPolicy1” : {},
+    “UnknownPolicy2” : {}
+  ]
+}
+```
+
+Examples of valid service configs :
+* Unknown fields are allowed
+```
+“serviceConfig” : {
+  “UnknownField” : “value”, // Field is unknown and hence ignored
+  “methodConfig” : {}
+}
+```
+
+
+### Service Config Channel Arguments
+
+gRPC allows two channel arguments to set that affect the behavior of service
+configs. One argument provides a service config to use if the resolver does not
+return a service config, and the other provides an option to disable service
+config resolution.
+
+If a default service config is provided, the client should fallback on the
+default service config (in case of an invalid config with no valid service
+config received earlier). On the other hand, if there is no default service
+config provided as a channel argument, the client should fail to start and keep
+waiting for a valid service config. Note that the default service config
+being set to an empty service config would be a valid fallback.
+
+## Implementation
+
+To implement the above spec -
+1. The resolver needs to be provided a parsing method(s) that it can run for a
+service config. This function will be provided to the resolver at instantiation
+time along with the default service config (if registered). This method will
+return a parsed form of the service config, or an error indicating that parsing
+failed.
+2. When a resolver gets an update, it runs the parser on this service config.
+	1. If parsing succeeds, it saves this service config and returns it to
+        the channel to apply.
+	2. If parsing fails, it provides a previously saved service config
```

Comment:
In Java (and it may be a good idea for other languages), we're planning on giving the full parse results (parsed_object, error) to the channel. That would allow the channel to use the old config. (And based on my comment above, use the old addresses+old config.)

Reply:
view comment above this failing when new addresses are received with a bad service config and the entire system going down even when a default service config is provided.

Reply:
We should change this to say that if parsing fails, the resolver just returns the parse error alongside the addresses.